node/pkg/watchers/sui/watcher.go

```diff
@@ -503,25 +450,154 @@ func (e *Watcher) Run(ctx context.Context) error {
 
 	select {
 	case <-ctx.Done():
-		_ = ws.Close(websocket.StatusNormalClosure, "")
 		return ctx.Err()
 	case err := <-errC:
-		_ = ws.Close(websocket.StatusInternalError, err.Error())
 		return err
 	}
 }
 
-// fixSuiWsURL ensures the websocket scheme is properly specified
-func (e *Watcher) fixSuiWsURL(logger *zap.Logger) error {
-	u, _ := url.Parse(e.suiWS)
+func (w *Watcher) getEvents() (SuiEventResponseData, error) {
+	// Only get events newer than the last processed height
+	var retVal SuiEventResponseData
+	var nextCursor struct {
+		TxDigest string
+		EventSeq string
+	}
+	firstTime := true
+	done := false
+	for !done {
+		var reader io.Reader
+		if firstTime {
+			reader = strings.NewReader(
+				fmt.Sprintf(`{"jsonrpc":"2.0", "id": 1, "method": "suix_queryEvents", "params": [{ "MoveEventType": "%s" }, null, %d, %t]}`,
+					w.suiMoveEventType, w.maximumBatchSize, w.descendingOrder))
+		} else {
+			reader = strings.NewReader(
+				fmt.Sprintf(`{"jsonrpc":"2.0", "id": 1, "method": "suix_queryEvents", "params": [{ "MoveEventType": "%s" }, { "txDigest": "%s", "eventSeq": "%s" }, %d, %t]}`,
+					w.suiMoveEventType, nextCursor.TxDigest, nextCursor.EventSeq, w.maximumBatchSize, w.descendingOrder))
+		}
+		resp, err := http.Post(w.suiRPC, "application/json", reader) //nolint:noctx // TODO FIXME we should propagate context with Deadline here.
+		if err != nil {
+			return retVal, fmt.Errorf("suix_queryEvents failed to post: %w", err)
+		}
+		body, err := io.ReadAll(resp.Body)
+		if err != nil {
+			return retVal, fmt.Errorf("suix_queryEvents failed to read: %w", err)
+		}
+		resp.Body.Close()
+
+		var res SuiEventResponse
+		err = json.Unmarshal(body, &res)
+		if err != nil {
+			return retVal, fmt.Errorf("suix_queryEvents failed to unmarshal body: %s, error: %w", string(body), err)
+		}
+		fmt.Println("Number of events:", len(res.Result.Data))
+		for _, datum := range res.Result.Data {
+			// Check if the event is newer than the last processed height
+			height, hErr := w.getCheckpointForEvent(datum)
+			if hErr != nil {
+				fmt.Println("Error getting checkpoint for event:", hErr)
+				return retVal, hErr
+			}
+			fmt.Println("Comparing Height:", height, "LastProcessedHeight:", w.latestProcessedCheckpoint)
+			if height <= w.latestProcessedCheckpoint {
+				done = true
+				fmt.Println("Done processing events.")
+				break
+			} else {
+				fmt.Println("Adding event to retVal")
+				retVal.Data = append(retVal.Data, datum)
+			}
+		}
+		if (!done) && res.Result.HasNextPage {
+			fmt.Println("Getting next page...")
+			nextCursor.TxDigest = res.Result.NextCursor.TxDigest
+			nextCursor.EventSeq = res.Result.NextCursor.EventSeq
+		} else {
+			done = true
+		}
+	}
+
+	retVal.Data = reverseArray(retVal.Data)
+	return retVal, nil
+}
+
+func (e *Watcher) getLatestCheckpointSN(logger *zap.Logger) (int64, error) {
```

Comment:
Can this function be used by `sui_block_height` as well?

Reply:
Yes. Changed.